File: FeDiT/OSFW/HyMS_eins/sim_engine/projects/watertank/watertank.m
classdef watertank < CsModel
   properties (Constant)
       BIG_A_VALUE = 2;         % 물탱크 단면 너비
       SMALL_A_VALUE = 0.5;     % 나가는 물의 양에 대한 상수
       MAX_HEIGHT = 0.5;        % 물탱크 최대 높이
   end
   properties
      water_height;
      is_height_under;      % 물 높이가 최대제한 수위 보다 낮은지
      file_id;
      in_water=1.0;
      under_height;
   end
   methods
        function this = watertank()
            this.VERSION = 1.0;
            this.reset();
            this.MODEL_NAME = 'watertank';
            
            this.add_inport('water_src', 0.0,'');

            this.add_outport('height', this.water_height,'');
            this.add_outport('water_over_height', int32(0),'');
            this.add_outport('water_under_height', int32(0),'');
            

            this.under_height = this.MAX_HEIGHT*0.8;
            
        end
        
        function delete(this)
            
        end

        function reset(this)
            disp('reset');
            this.water_height = 0.0;
            this.is_height_under = 1;
            this.TA = 0.1;
            this.in_water = 0.0;
        end

        function ret = set_param(this, in_param)
            ret = true;
        end

        function ret = ExtTransFn(this, sim_time, dt, in_port)
            this.in_water = in_port.value;
            
            this.set_continue();
            ret = true;
        end
        
        function ret = func(this, t, y)
            if this.water_height >= 0.0
                ret = (this.in_water - sqrt(y) * this.SMALL_A_VALUE) / this.BIG_A_VALUE;
            else
                ret = this.in_water / this.BIG_A_VALUE;
            end
        end
                
        function ret = IntTransFn(this, sim_time, dt)
            this.water_height = RK4(@this.func, this.water_height, sim_time, dt);
            ret = true;
        end 
        
        function ret = OutputFn(this, sim_time, dt)
            %fprintf(this.file_id, '%f,%f,%f\n', sim_time, this.water_height, this.in_water);
            if this.water_height > this.MAX_HEIGHT && this.is_height_under == 1
               this.is_height_under = 0;
               this.set_outport_value('water_over_height', int32(1), '');
            elseif this.water_height < this.under_height && this.is_height_under == 0
               this.is_height_under = 1;
               this.set_outport_value('water_under_height', int32(1), '');
            end
            this.set_outport_value('height', this.water_height, '');
            ret = true;
        end
        
        function ret = TimeAdvanceFn(this)
            ret = 0.1;
        end
           
   end
 
end

File: FeDiT/OSFW/HyMS_eins/sim_engine/projects/watertank/watertank.py
from enum import Enum
import json
from scipy import integrate as inte
import math
from CsModel import CsModel 

BIG_A_VALUE = 2.0       # 물탱크 단면 너비
SMALL_A_VALUE = 0.5     # 나가는 물의양에 대한 상수
MAX_HEIGHT = 0.5        # 물탱크 최대 높이


class watertank(CsModel):
    def __init__(self):
        self.MODEL_NAME = "watertank"
	 
        self.add_inport('water_src', 0.0)
        self.add_outport('water_over_height', 0)
        self.add_outport('water_under_height', 0)
        
        self.VERSION = 1.0
        self.reset()
        #self.log_fp = open('watertank_log_py.txt', 'w')
  
    def __del__(self):
        #self.log_fp.close()
        self.state = 1
 
    def reset(self):
        self.water_height = 0.0
        self.is_height_under = True
        self.in_water = 0.0
        self.under_height = MAX_HEIGHT * 0.8
        self.state = 1

    def rk4(self, fun, y, t, h):
        k1 = fun(t,y)
        k2 = fun(t+1/2*h,y+1/2*h*k1)
        k3 = fun(t+1/2*h,y+1/2*h*k2)
        k4 = fun(t+h,y+h*k3)

        y = y + 1/6*h*(k1 + 2*k2 + 2*k3 + k4)
        t = t + h
        return y

    def func(self, t, y):
        if self.water_height >= 0.0:
            return (self.in_water - math.sqrt(y) * SMALL_A_VALUE) / BIG_A_VALUE
        else:
            return self.in_water / BIG_A_VALUE

    def ExtTransFn(self, sim_time, dt, src_model_name, port_name, port_value):
        self.in_water = port_value
        """
        sol = inte.RK45(self.func, sim_time, [self.water_height], sim_time + 0.1, 1)
        
        for i in range(100):
            sol.step()
            #print(i, sol.t, sol.y[0])
            if sol.status == 'finished':
                break
        self.water_height = sol.y[0]    
        """
        
        #print(self.water_height)
        #self.set_continue()
        return True    

    def IntTransFn(self, sim_time, dt):
        self.water_height = self.rk4(self.func, self.water_height, sim_time, dt)
        return True

    def OutputFn(self, sim_time, dt):
        #self.log_fp.write(f'{sim_time},{self.water_height},{self.in_water}\n')
        #self.log_fp.flush()
        if self.water_height > MAX_HEIGHT and self.is_height_under == True:
            self.is_height_under = False
            self.set_outport_value('water_over_height', 1)
            print("Approx Tank Over Height : ", self.water_height, " at t = ", sim_time)
        elif self.water_height < self.under_height and self.is_height_under == False:
            self.is_height_under = True
            self.set_outport_value('water_under_height', 1)
            print("Approx Tank Under Height : ", self.water_height, " at t = ", sim_time)
        else:
            print("ApproxTank Height : ", self.water_height, " at t = ", sim_time)
        return True

    def TimeAdvanceFn(self, sim_time):
        return 0.1


if __name__ == "__main__":
    w = watertank()
    w.ExtTransFn(0.1, 'water_src', 1.0)
    print(w.MODEL_NAME, w.water_height)


   

